src/frontends/onnx/frontend/src/op/scan.cpp

```diff
@@ -0,0 +1,176 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "op/scan.hpp"
+
+#include <iterator>
+#include <memory>
+
+#include "core/graph.hpp"
+#include "default_opset.hpp"
+#include "exceptions.hpp"
+#include "ngraph/function.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/op/util/op_types.hpp"
+#include "onnx_import/core/null_node.hpp"
+#include "openvino/core/validation_util.hpp"
+
+namespace ngraph {
+namespace onnx_import {
+namespace op {
+
+namespace {
+
+OutputVector scan_to_tensor_iterator(const OutputVector& node_inputs,
+                                     ParameterVector& body_inputs,
+                                     OutputVector& body_outputs,
+                                     int64_t num_scan_inputs,
+                                     const std::vector<int64_t>& scan_input_axes,
+                                     const std::vector<int64_t>& scan_input_directions,
+                                     const std::vector<int64_t>& scan_output_axes,
+                                     const std::vector<int64_t>& scan_output_directions,
+                                     int64_t in_offset = 0,
+                                     const std::string& node_description = "") {
+    const size_t num_initial_values = body_inputs.size() - num_scan_inputs;
+    const size_t num_scan_outputs = body_outputs.size() - num_initial_values;
+
+    // Body inputs alignment
+    for (size_t i = 0; i < num_initial_values; ++i) {
+        body_inputs[i]->set_element_type(node_inputs[i + in_offset].get_element_type());
+        body_inputs[i]->set_partial_shape(node_inputs[i + in_offset].get_partial_shape());
+        body_inputs[i]->validate_and_infer_types();
+    }
+    // Single slice of TensorIterator sliced input has the same rank as the input,
+    // but in ONNX Scan the slice of input can has one dimension less,
+    // so the parameter needs to have aligned rank with 1 at sliced axis,
+    // and then squeezed to restore original shape.
+    for (size_t i = 0; i < num_scan_inputs; ++i) {
+        const auto in_idx = num_initial_values + i;
+        const auto axis = ov::normalize_axis(node_description,
+                                             scan_input_axes[i],
+                                             node_inputs[in_idx + in_offset].get_partial_shape().rank());
+        const auto axis_node = default_opset::Constant::create(element::i64, Shape{1}, {axis});
```

Comment:
Why do we normalize it? (Un)Squeeze supports negative indices. But in case of dynamic rank we would have an error.

Reply:
Later we need to throw on `normalize_axis` for such case anyway, because negative axis indices are not supported for `SliceInputDescription` (set_sliced_input) or `ConcatOutputDescription` (get_concatenated_slices) in TensorIterator validate_and_infer_types.
But ok, here to fix body_input Parameter shape, the axis normalization can be applied only for static rank.